addons/mail/models/mail_message.py

```diff
@@ -988,11 +994,20 @@ def _message_format(self, fnames):
                         'field_type': tracking.field_type,
                     })
 
+            if message_sudo.model and message_sudo.res_id:
+                record_name = self.env[message_sudo.model] \
+                    .browse(message_sudo.res_id) \
+                    .with_prefetch(thread_ids_by_model_name[message_sudo.model]) \
+                    .display_name
```

Comment:
Actually I was hoping to merge but I realize we're lacking a `sudo()` here again (it was previously on the manual prefetch before all the recent changes).

We really need to introduce a test to cover this case, I'm pretty sure this can fail in at least 2 situation, being mentioned in a record that is not available due to:
- access rules on the record
- or record belonging to another company in a multi-company setup

Reply:
I have added the `sudo`.
@tde-banana-odoo also wanted to test `record_name` (https://github.com/odoo/odoo/pull/63000#discussion_r586225521), I guess we should test:

- That the `record_name` match the `thread_name`
- That the `record_name` is updated when the `thread_name` is updated ? (this one sound useless)
- We have `record_name` even if we don't have access to the record
- Multi-company ?

Reply:
Point 1 you have done it already in "test message format".
Point 2 is technically useless to test due to implementation but it is functionally desirable, so we could test it.
Point 3+4 yes that's the one I'm requesting (and we actually need to decide from a functional and security standpoint if we allow the name to leak in that case, or if we want to handle the exceptions and have no name)